src/utils/listener/block_downloader.go

```diff
@@ -226,3 +198,124 @@ func (self *BlockDownloader) run() error {
 
 	return nil
 }
+
+func (self *BlockDownloader) downloadBlocks(height uint64, lastProcessedBlockHash arweave.Base64String) (out []*arweave.Block, err error) {
+	var mtx sync.Mutex
+	var wg sync.WaitGroup
+
+	peers := append(self.client.GetCachedPeers(), self.Config.Arweave.NodeUrl)
+	out = make([]*arweave.Block, 0, len(peers))
+	wg.Add(len(peers))
+
+	// Download blocks from all peers in parallel
+	for _, peer := range peers {
+		peer := peer
+		self.SubmitToWorker(func() {
+			block, err := self.downloadOneBlock(height, lastProcessedBlockHash, peer)
+			if err != nil {
+				block = nil
+			}
+			mtx.Lock()
+			if block != nil {
+				out = append(out, block)
+			}
+			mtx.Unlock()
+			wg.Done()
+		})
+	}
+
+	// Wait for all workers to finish
+	wg.Wait()
+
+	return
+}
+
+func (self *BlockDownloader) downloadOneBlock(height uint64, lastProcessedBlockHash arweave.Base64String, peer string) (block *arweave.Block, err error) {
+	ctx := context.WithValue(self.Ctx, arweave.ContextDisablePeers, true)
+	if len(peer) > 0 {
+		// Force using this peer if set
+		// Otherwise use arweave.net
+		ctx = context.WithValue(ctx, arweave.ContextForcePeer, peer)
+	}
+
+	block, resp, err := self.client.GetBlockByHeight(ctx, int64(height))
```

Comment:
Does this function loads full block (including its data) or only the headers?

For the voting stage I believe that it should load only the headers

Reply:
...and load the rest data after the 'right' block hash has been chosen.